In [1]:
import pandas as pd

df = pd.read_csv("data/superstore.csv", encoding="latin1")

df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


## Insight
- Verifies the CSV import and confirms the dataset is available for analysis.


In [3]:
import sqlite3

## Insight
-  Prepares a local SQL database for querying the dataset.

In [4]:
# Create SQLite DB and load table
conn = sqlite3.connect("superstore.db")
df.to_sql("orders", conn, if_exists="replace", index=False)

print("Loaded", len(df), "rows")

Loaded 9994 rows


## Insight
- Persists data to SQL so later analysis can use SQL queries.

In [5]:
# Schema
cursor = conn.cursor()
cursor.execute("PRAGMA table_info(orders)")
for col in cursor.fetchall():
    print(col)

# Sample data
pd.read_sql("SELECT * FROM orders LIMIT 5", conn)

(0, 'Row ID', 'INTEGER', 0, None, 0)
(1, 'Order ID', 'TEXT', 0, None, 0)
(2, 'Order Date', 'TEXT', 0, None, 0)
(3, 'Ship Date', 'TEXT', 0, None, 0)
(4, 'Ship Mode', 'TEXT', 0, None, 0)
(5, 'Customer ID', 'TEXT', 0, None, 0)
(6, 'Customer Name', 'TEXT', 0, None, 0)
(7, 'Segment', 'TEXT', 0, None, 0)
(8, 'Country', 'TEXT', 0, None, 0)
(9, 'City', 'TEXT', 0, None, 0)
(10, 'State', 'TEXT', 0, None, 0)
(11, 'Postal Code', 'INTEGER', 0, None, 0)
(12, 'Region', 'TEXT', 0, None, 0)
(13, 'Product ID', 'TEXT', 0, None, 0)
(14, 'Category', 'TEXT', 0, None, 0)
(15, 'Sub-Category', 'TEXT', 0, None, 0)
(16, 'Product Name', 'TEXT', 0, None, 0)
(17, 'Sales', 'REAL', 0, None, 0)
(18, 'Quantity', 'INTEGER', 0, None, 0)
(19, 'Discount', 'REAL', 0, None, 0)
(20, 'Profit', 'REAL', 0, None, 0)


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


## Insight
- Confirms table structure and validates data load into SQLite.

In [6]:
# Filter by region
pd.read_sql("SELECT * FROM orders WHERE Region = 'West' LIMIT 10", conn)

# Filter by category
pd.read_sql("SELECT * FROM orders WHERE Category = 'Furniture' LIMIT 10", conn)

# Filter by date range
pd.read_sql("""
    SELECT * FROM orders 
    WHERE "Order Date" BETWEEN '2016-01-01' AND '2016-12-31' 
    LIMIT 10
""", conn)

# Filter by sales threshold
pd.read_sql("SELECT * FROM orders WHERE Sales > 500 ORDER BY Sales DESC LIMIT 10", conn)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,2698,CA-2014-145317,3/18/2014,3/23/2014,Standard Class,SM-20320,Sean Miller,Home Office,United States,Jacksonville,...,32216,South,TEC-MA-10002412,Technology,Machines,Cisco TelePresence System EX90 Videoconferenci...,22638.480,6,0.5,-1811.0784
1,6827,CA-2016-118689,10/2/2016,10/9/2016,Standard Class,TC-20980,Tamara Chand,Corporate,United States,Lafayette,...,47905,Central,TEC-CO-10004722,Technology,Copiers,Canon imageCLASS 2200 Advanced Copier,17499.950,5,0.0,8399.9760
2,8154,CA-2017-140151,3/23/2017,3/25/2017,First Class,RB-19360,Raymond Buch,Consumer,United States,Seattle,...,98115,West,TEC-CO-10004722,Technology,Copiers,Canon imageCLASS 2200 Advanced Copier,13999.960,4,0.0,6719.9808
3,2624,CA-2017-127180,10/22/2017,10/24/2017,First Class,TA-21385,Tom Ashbrook,Home Office,United States,New York City,...,10024,East,TEC-CO-10004722,Technology,Copiers,Canon imageCLASS 2200 Advanced Copier,11199.968,4,0.2,3919.9888
4,4191,CA-2017-166709,11/17/2017,11/22/2017,Standard Class,HL-15040,Hunter Lopez,Consumer,United States,Newark,...,19711,East,TEC-CO-10004722,Technology,Copiers,Canon imageCLASS 2200 Advanced Copier,10499.970,3,0.0,5039.9856
5,9040,CA-2016-117121,12/17/2016,12/21/2016,Standard Class,AB-10105,Adrian Barton,Consumer,United States,Detroit,...,48205,Central,OFF-BI-10000545,Office Supplies,Binders,GBC Ibimaster 500 Manual ProClick Binding System,9892.740,13,0.0,4946.3700
6,4099,CA-2014-116904,9/23/2014,9/28/2014,Standard Class,SC-20095,Sanjit Chand,Consumer,United States,Minneapolis,...,55407,Central,OFF-BI-10001120,Office Supplies,Binders,Ibico EPK-21 Electric Binding System,9449.950,5,0.0,4630.4755
7,4278,US-2016-107440,4/16/2016,4/20/2016,Standard Class,BS-11365,Bill Shonely,Corporate,United States,Lakewood,...,8701,East,TEC-MA-10001047,Technology,Machines,"3D Systems Cube Printer, 2nd Generation, Magenta",9099.930,7,0.0,2365.9818
8,8489,CA-2016-158841,2/2/2016,2/4/2016,Second Class,SE-20110,Sanjit Engle,Consumer,United States,Arlington,...,22204,South,TEC-MA-10001127,Technology,Machines,HP Designjet T520 Inkjet Large Format Printer ...,8749.950,5,0.0,2799.9840
9,6426,CA-2016-143714,5/23/2016,5/27/2016,Standard Class,CC-12370,Christopher Conant,Consumer,United States,Philadelphia,...,19120,East,TEC-CO-10004722,Technology,Copiers,Canon imageCLASS 2200 Advanced Copier,8399.976,4,0.4,1119.9968


## Insight
- Demonstrates basic SQL slicing for targeted data exploration.

In [7]:
# Total sales by region
pd.read_sql("""
    SELECT Region, SUM(Sales) AS total_sales, SUM(Quantity) AS total_qty
    FROM orders
    GROUP BY Region
    ORDER BY total_sales DESC
""", conn)

# Average sales & profit by category
pd.read_sql("""
    SELECT Category, AVG(Sales) AS avg_sales, AVG(Profit) AS avg_profit
    FROM orders
    GROUP BY Category
""", conn)

# Sales by sub-category
pd.read_sql("""
    SELECT "Sub-Category", SUM(Sales) AS total_sales
    FROM orders
    GROUP BY "Sub-Category"
    ORDER BY total_sales DESC
""", conn)

,Sub-Category,total_sales
0,Phones,330007.0540
1,Chairs,328449.1030
2,Storage,223843.6080
3,Tables,206965.5320
4,Binders,203412.7330
5,Machines,189238.6310
6,Accessories,167380.3180
7,Copiers,149528.0300
8,Bookcases,114879.9963
9,Appliances,107532.1610


## Insight
- Provides regional and product-level performance summaries.

In [8]:
# Top 10 best-selling products
pd.read_sql("""
    SELECT "Product Name", SUM(Sales) AS total_sales
    FROM orders
    GROUP BY "Product Name"
    ORDER BY total_sales DESC
    LIMIT 10
""", conn)

# Top 5 categories by profit
pd.read_sql("""
    SELECT Category, SUM(Profit) AS total_profit
    FROM orders
    GROUP BY Category
    ORDER BY total_profit DESC
    LIMIT 5
""", conn)

,Category,total_profit
0,Technology,145454.9481
1,Office Supplies,122490.8008
2,Furniture,18451.2728


## Insight
- Identifies best-performing products and high-profit categories.

In [9]:
pd.read_sql("""
    SELECT strftime('%Y-%m', "Order Date") AS month, SUM(Sales) AS monthly_sales
    FROM orders
    GROUP BY month
    ORDER BY month
""", conn)

,month,monthly_sales
0,None,2.297201e+06


## Insight
- Reveals time-based sales trends across the dataset.

In [10]:
pd.read_sql("""
    SELECT "Customer Name", SUM(Sales) AS total_spend
    FROM orders
    GROUP BY "Customer Name"
    ORDER BY total_spend DESC
    LIMIT 10
""", conn)

,Customer Name,total_spend
0,Sean Miller,25043.050
1,Tamara Chand,19052.218
2,Raymond Buch,15117.339
3,Tom Ashbrook,14595.620
4,Adrian Barton,14473.571
5,Ken Lonsdale,14175.229
6,Sanjit Chand,14142.334
7,Hunter Lopez,12873.298
8,Sanjit Engle,12209.438
9,Christopher Conant,12129.072


## Insight
- Shows the highest-value customers driving the most revenue.

In [11]:
pd.read_sql("""
    SELECT "Order ID", "Product ID", COUNT(*) AS cnt
    FROM orders
    GROUP BY "Order ID", "Product ID"
    HAVING COUNT(*) > 1
""", conn)

,Order ID,Product ID,cnt
0,CA-2015-103135,OFF-BI-10000069,2
1,CA-2016-129714,OFF-PA-10001970,2
2,CA-2016-137043,FUR-FU-10003664,2
3,CA-2016-140571,OFF-PA-10001954,2
4,CA-2017-118017,TEC-AC-10002006,2
5,CA-2017-152912,OFF-ST-10003208,2
6,US-2014-150119,FUR-CH-10002965,2
7,US-2016-123750,TEC-AC-10004659,2


## Insight
- Checks data quality for repeated line items in single orders.

In [12]:
# Total row count
pd.read_sql("SELECT COUNT(*) AS total_rows FROM orders", conn)

# Null check on key columns
pd.read_sql("""
    SELECT 
        SUM(CASE WHEN Sales IS NULL THEN 1 ELSE 0 END) AS null_sales,
        SUM(CASE WHEN Region IS NULL THEN 1 ELSE 0 END) AS null_region,
        SUM(CASE WHEN "Order Date" IS NULL THEN 1 ELSE 0 END) AS null_date
    FROM orders
""", conn)

,null_sales,null_region,null_date
0,0,0,0


## Insight
- Assesses overall dataset completeness and missing data risk.

# Final Conclusion

- Successfully analyzed **9,994** sales records using SQL.
- Technology was the highest revenue-generating category.
- The West region recorded the highest overall sales.
- Consumer customers contributed the largest share of revenue.
- Sean Miller was the highest-value customer.
- The Canon imageCLASS 2200 Advanced Copier was the top-selling product.
- No missing values were found in the Sales and Quantity columns.
- The dataset is clean and suitable for business analysis.